### ISSUE TITLE
Flexible Zeitraumauswahl

### ISSUE THREAD

**Issue by bennysweb** [solectrus/solectrus#1931]
Die Auswahl des darzustellenden Zeitraums soll flexibler erfolgen:

- Datepicker zum Auswählen eines Jahres, eines Monats, einer Woche, eines Tages (siehe auch #1931). Das erspart das umständliche Blättern zu einem gewünschten Zeitraum

- Ermöglichen beliebiger Zeiträume durch Angabe von Start- und EndDatum (siehe #3004)

- Flexible Gruppierung im Diagramm: Ein großer Zeitraum sollte mit Balken für jeden Monat erfolgen, bei kleineren eher Tage oder Wochen. Denkbar wäre auch, die Gruppierung vom Benutzer festlegen zu lassen. Bei sehr kleinen Zeiträumen (weniger als sieben Tage) sollte gar keine Gruppierung erfolgen, d.h. ein Liniendiagramm benutzten werden, siehe #4033)

- Hat man einen größeren Zeitraum gewählt, der nicht nur vollständige Monate oder Wochen umfasst (z.B. 10.09.2023 bis 27.03.2024), dann wäre der erste und/und der letzte Balken kein vollständiger Monat bzw. Woche. Das könnte entsprechend visualisiert werden, z.b. durch schraffierte Darstellung o.ä.

- Relative Zeiträume wie z.B. die "letzten 7 Tage" oder "die letzten 12 Monate"

--- 
Ursprünglicher Beitrag von @bennysweb:

> Hi Georg,
> just an idea: Can you implement a calendar icon on the top right corner (left to the menue > button) to select a day, month or year to jump to it?
> Now we have to click very often to find a specific date and to look at the values.

**Comment by Zeppelin-PV** [comment]
I agree, bennysweb
A calender function would bei helpfull
It should have also an hour selection eg. from... to....
to find a specific time range

**Comment by asieberle** [comment]
Also support this requests, especially hourly selection would bei very helpful to identify/verify high power consumers

**Comment by halmich** [comment]
Yes, I also like the idea of having a day (calendar?) and month picker.

**Comment by bennysweb** [comment]
Hi Georg, after 2 years with Solectrus and a lot of days/months I want to compare and check out again I just want to put this back on the table. Is there any motivation on your side to implement the calender picker?

**Comment by ledermann** [comment]
A date picker is undoubtedly a useful feature that is still on my to-do list. But there are also many other things on that list. 

We also need a picker that works for days, weeks, months and years. There are many solutions out there, I need to choose a good one.

**Comment by halmich** [comment]
ich Vergleiche tatsächlich gerne den Verbrauch oder auch den Sonnenertrag eines Monats mit dem Monat der Vorjahre.
fände da ein Button cool, dass man von Januar 2024 direkt zu Januar 2023 und dann zu Januar 2022 springen könnte. Und zurück 😉

**Comment by bennysweb** [comment]
I'm absolutely with you @ledermann . Maybe the community can support and push this feature with some Ko-fi donations?

**Comment by ledermann** [comment]
Einen kleinen Teil habe ich hier nun realisiert, allerdings noch **ohne Benutzeroberfläche**.

Per Manipulation der URL lassen sich jetzt relative Zeiträume darstellen, indem der gewünschte Zeitraum in **Anlehnung** an [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601#Durations) definiert wird (derzeitig nur für Tage und Monate implementiert).

Zwei Beispiele:

- Die letzten 30 Tage: 
  https://demo.solectrus.de/inverter_power/P30D
- Die letzten 12 Monate: 
  https://demo.solectrus.de/inverter_power/P12M

Zu sehen in `develop` ab `v0.18.3-404` (für Sponsoren).

**Comment by ledermann** [comment]
Wer als Sponsor auf `develop` ist und die `v0.19.0-24` hat, der kann dort sehen, dass die Buttons "Woche", "Monat" und "Jahr" eine Doppelfunktion bekommen haben. Ein erneuter Klick führt zu einer etwas anderen Darstellung, und zwar:

- Aktuelle Woche → Letzte 7 Tage
- Aktueller Monat → Letzte 30 Tage
- Aktuelles Jahr → Letzte 12 Monate

Das ist noch unausgereift, aber hat schon einen gewissen Mehrwert. Ich arbeite an weiteren Auswahlmöglichkeiten.

**Comment by Zeppelin-PV** [comment]
das ist eine gute Lösung, deckt schon mal ein Großteil der Recherchen ab.
Ist das für den Tag nicht auch noch möglich ?
Tag / letzte 24h

**Comment by ledermann** [comment]
@Zeppelin-PV Da die Tageszusammenfassungen die wesentliche Grundlage für die Berechnung der Strombilanz sind und prinzipiell immer **vollständige** Tage umfassen, müssen sämtliche Darstellungen immer um 00:00 Uhr beginnen. So etwas wie "die letzten 24h" wird also nicht möglich sein. 

Was aber gehen wird, ist z.B. "die letzten drei Tage" - dann soll (demnächst) ein Liniendiagramm erscheinen, also keine Säulen. Man würde dann z.B. bei Erzeugung **mehrere** Glockenkurven nebeneinander sehen.

**Comment by hefe193** [comment]
Es ist doch aktuell auch möglich aus der Jahresansicht einen Monat im Diagramm anzuklicken, dann kommt man zu der Monatsansicht des Monats. Dann dort auf einen Tag geklickt sehe ich die Grafiken und Details dieses Tages.

**Comment by ledermann** [comment]
@hefe193 Ja, das [Drilldown](https://de.wikipedia.org/wiki/Drilldown) gibt es schon länger (seit [v0.17.1](https://github.com/solectrus/solectrus/releases/tag/v0.17.1)): Man kann sich vom Jahr bis zum Tag durchklicken. Hier geht es aber darum, beliebige Zeiträume darzustellen, also nicht nur einzelne Tage/Wochen/Monate/Jahre, sondern Zeiträume von A bis B oder die letzten X Tage/Monate/Jahre.

**Comment by ledermann** [comment]
Auch die letzten 24h sind jetzt darstellbar (angesprochen von @Zeppelin-PV und in #4424). Anstelle der Tageszusammenfassungen wird die Summe dafür direkt aus InfluxDB berechnet, das ist bei der kurzen Zeitspanne vertretbar.

Damit kann in der Hauptnavigation der Menüpunkt "Tag" auch als Toggle verwendet werden. Sieht man gerade den aktuellen Tag, führt ein erneuter Klick zu den letzten 24 Stunden:
https://demo.solectrus.de/inverter_power/P24H

Durch manuelles Verändern der URL kann man auch andere Abschnitte darstellen (maximal 99h), z.B. die letzten 72 Stunden (= 3 Tage als Liniendiagramm):
https://demo.solectrus.de/inverter_power/P72H

Eine vernünftige Zeitraumauswahl kommt erst noch. Bis dahin ist das ein Insider-Feature ;-)
